### PR TITLE
Refactor KNNCodec to use new extension point

### LINKED ISSUE #146: [BUG] KNNCodecService is incompatible with Lucene best_compression Codec Mode
The current [KNN87Codec](https://github.com/opensearch-project/k-NN/blob/d7c55e9bdbc37d71a714569c9f37755053781c7d/src/main/java/org/opensearch/knn/index/codec/KNN87Codec/KNN87Codec.java#L81) uses a Delegate pattern for all Field Formats except `DocValues` and `CompoundFormat`. The problem with this delegate is that it always returns the default `LUCENE_87` format loaded through SPI making it incompatible with `best_compression` option available through the [index_settings](https://opensearch.org/docs/latest/opensearch/rest-api/index-apis/create-index/#index-settings) configuration.

What this means to the end user is a misleading / unexpected behavior in the following situation:

```
"settings" : {
  "index" : {
    "codec" : "best_compression",
    "knn" : "true"
  }
} 
```

When `knn` is set to `true` the [KNNCodecService](https://github.com/opensearch-project/k-NN/blob/main/src/main/java/org/opensearch/knn/plugin/KNNCodecService.java) is hot swapped in place of the default CodecService. As seen in the code the KNNCodecService [always returns](https://github.com/opensearch-project/k-NN/blob/d7c55e9bdbc37d71a714569c9f37755053781c7d/src/main/java/org/opensearch/knn/plugin/KNNCodecService.java#L52) `KNN87Codec.KNN_87` which uses the delegate pattern as described above. This means `index.codec` parameter value will be silently ignored  since the `name` is ignored in the `codec(String name)` overridden method.

Users will think they get `best_compression` on their fields but, sike, they won't. This is not a good experience and should be fixed so that (at minimum) the `best_compression` mode can be passed to the `KNNCodecService` giving the users the option to use compression on any fields other than compound and docValues formats. 

Documentation should also be updated to provide full transparency of codec behavior.

### PR DESCRIPTION
**Pull request by jmazanec15**
### Description
This is a fairly large refactor of our current KNNCodec system.

First, it deletes unused code. For instance, we had KNN84Codec and KNN86Codec that we were not using. The reason to keep them had been potential assistance with backwards compatibility. However, given that we do not use them now, I think it makes sense to just delete.

Second, the plugin is refactored to implement getCustomCodecServiceFactory. This is a new extension point in OpenSearch and will allow multiple EnginePlugins to be used on the same index. For instance, the CCR plugin and k-NN. This allows us to delete the EngineFactory code. Along with this, I moved KNNCodecService to the codec folder for consistency.

Third, the KNN87Codec was refactored to implement the FilterCodec. The KNN87Codec uses the delegate pattern to override the DocValuesFormat. The rest of the formats are delegated to a delegate. This is what the FilterCodec is intended to be used for. Along with this, I extended the delegate reach further so that the KNN87Codec could implement the pattern to its fullest potential.

Fourth, I added several unit tests for the individual components of the KNN87Codec as well as a util class. This will allow us to cover to a fuller extent the test coverage around these components. I have not yet added unit test coverage for merging or for the KNNCodecUtil. I will do this in the future.

**Note** - I expect checkstyle to fail. My plan is to run `./gradlew spotlessApply` after I get review so as to not distract reviewers from the substantive changes.
 
### Issues Resolved
#146 #147 
 
### Check List
- [X] New functionality includes testing.
  - [X] All tests pass
- [X] New functionality has been documented.
  - [X] New functionality has javadoc added
- [X] Commits are signed as per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/k-NN/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/main/java/org/opensearch/knn/index/codec/KNN80Codec/KNN80DocValuesConsumer.java
Comment (VijayanB) [inline src/main/java/org/opensearch/knn/index/codec/KNN80Codec/KNN80DocValuesConsumer.java:93]: Do we need null check for attributes before calling contains?

File: src/main/java/org/opensearch/knn/index/codec/KNNCodecService.java
Comment (martin-gaievski) [inline src/main/java/org/opensearch/knn/index/codec/KNNCodecService.java:36]: I think we need some sort of factory here that will allow to select exact k-NN Codec implementation. I'm thinking of a use case for 2.0/lucene 9, we'll need to use Lucene90 codec for it, and probably will create new kNN wrapper class, something like KNN90Codec. Although it might be too much for this PR, we can do it in next one.

File: src/main/resources/META-INF/services/org.apache.lucene.codecs.Codec
Comment (vamshin) [inline src/main/resources/META-INF/services/org.apache.lucene.codecs.Codec:3]: Are we sure we do not use these codecs in any older opendistro version? If these codecs were used to create segments in some older version then assuming they upgrade to latest OpenSearch version, it might fail as they do not find respective codec for reading.

Comment (jmazanec15) [reply src/main/resources/META-INF/services/org.apache.lucene.codecs.Codec:3]: Good catch. I think this could have problems for SPI. 

Let me add back these Codecs back.

File: src/main/java/org/opensearch/knn/index/codec/KNNCodecService.java
Comment (martin-gaievski) [reply src/main/java/org/opensearch/knn/index/codec/KNNCodecService.java:30]: Ok, I see, so essentially we change it with every next Lucene version but keep older codecs for bwc. Seems that this method (codec(String name)) is the factory method itself. Do we have our abstraction for knn codec apart from lucene abstract Codec class? It may give us more flexibility with testing, although I'm not sure it's that easy to pass our interface to the OpenSearch through plugin related service classes.

Comment (jmazanec15) [reply src/main/java/org/opensearch/knn/index/codec/KNNCodecService.java:30]: Right, so we keep older Codecs so that older indices can be read by old Codecs via SPI. 

BWC is also why we need to version our Codecs (KNN87Codec). This is so that older indices can be read with old codecs.

I think this is something that could be investigated more in a future review.

**Comment by jmazanec15** [comment]
@martin-gaievski @vamshin Failure related to spotless check. Please review again - Ill fix spotless after I get approval

File: src/main/java/org/opensearch/knn/index/codec/KNN87Codec/KNN87Codec.java
Comment (vamshin) [inline src/main/java/org/opensearch/knn/index/codec/KNN87Codec/KNN87Codec.java:52]: Are we not doing this for every code >= Lucene87? Wondering why this should be only specific to Lucene87

Comment (jmazanec15) [reply src/main/java/org/opensearch/knn/index/codec/KNN87Codec/KNN87Codec.java:52]: Right, so Codec does not have the method `getDocValuesFormatForField`. Lucene8XCodec's do, however. So, I felt like it made sense to map this to the LuceneCodec version our KNN codec version maps to.

Comment (martin-gaievski) [reply src/main/java/org/opensearch/knn/index/codec/KNN87Codec/KNN87Codec.java:52]: We can pull this to the separate DocValuesFormat class and inject to knn codec during construction time. We can also make getDocValuesFormatForField part of our codec facade abstraction, this should help to decrease coupling between our and lucene classes. These ideas are more for next refactoring.

Comment (jmazanec15) [reply src/main/java/org/opensearch/knn/index/codec/KNN87Codec/KNN87Codec.java:52]: @martin-gaievski what would this look like?

Comment (jmazanec15) [reply src/main/java/org/opensearch/knn/index/codec/KNN87Codec/KNN87Codec.java]: I actually was able to remove this code. The reason beingthat the delegate could have the per field functionality and we dont need to do anything to get it.

Comment (martin-gaievski) [reply src/main/java/org/opensearch/knn/index/codec/KNN87Codec/KNN87Codec.java]: nice. My initial idea was to create a facade interface for knn codec and then inject it to KNNCodecService using DI similar to spring or guice. I may try to play with it while working on integration with opensearch 2.0/lucene 9.1, we need a codec based on lucene 9.1

**Review by martin-gaievski — DISMISSED** [review]
lgtm

**Review by vamshin — DISMISSED** [review]
LGTM! Thanks

**Review by vamshin — APPROVED** [review]
LGTM! Thanks

**Review by martin-gaievski — APPROVED** [review]

**Pull request merged**
